jetbrains-core/tst/software/aws/toolkits/jetbrains/services/s3/S3ServiceNodeTest.kt

```diff
@@ -0,0 +1,97 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package software.aws.toolkits.jetbrains.services.s3
+
+import com.intellij.testFramework.ProjectRule
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import software.amazon.awssdk.services.s3.S3Client
+import software.amazon.awssdk.services.s3.model.Bucket
+import software.aws.toolkits.core.region.AwsRegion
+import software.aws.toolkits.jetbrains.core.MockClientManagerRule
+import software.aws.toolkits.jetbrains.core.MockResourceCache
+import software.aws.toolkits.jetbrains.core.credentials.MockProjectAccountSettingsManager
+import software.aws.toolkits.jetbrains.core.credentials.ProjectAccountSettingsManager
+import software.aws.toolkits.jetbrains.core.explorer.nodes.AwsExplorerEmptyNode
+import software.aws.toolkits.jetbrains.core.explorer.nodes.AwsExplorerErrorNode
+import software.aws.toolkits.jetbrains.services.s3.resources.S3Resources
+import software.aws.toolkits.jetbrains.utils.delegateMock
+import java.time.Instant
+import java.util.concurrent.CompletableFuture
+
+class S3ServiceNodeTest {
+
+    @JvmField
+    @Rule
+    val projectRule = ProjectRule()
+
+    @Rule
+    @JvmField
+    val mockClientManager = MockClientManagerRule { projectRule.project }
+
+    private val mockClient = delegateMock<S3Client>()
+    private val mockSettingsManager by lazy {
+        ProjectAccountSettingsManager.getInstance(projectRule.project)
+                as MockProjectAccountSettingsManager
+    }
+
+    @Before
+    fun setUp() {
+        resourceCache().clear()
+        mockSettingsManager.changeRegion(AwsRegion.GLOBAL)
+        mockClientManager.manager().register(S3Client::class, mockClient)
+    }
+
+    @Test
+    fun s3BucketsAreListed() {
+        val bucketList = listOf("bcd", "abc", "AEF", "ZZZ")
+        resourceCache().s3buckets(bucketList)
+        bucketList.map { resourceCache().bucketRegion(it) }
+        val children = S3ServiceNode(projectRule.project).children
+
+        assertThat(children).allMatch { it is S3BucketNode }
+        assertThat(children.filterIsInstance<S3BucketNode>().map { it.displayName() }).containsExactlyInAnyOrder(
+            "abc",
+            "AEF",
+            "bcd",
+            "ZZZ"
+        )
+    }
+
+    @Test
+    fun noBucketsInTheRegion() {
+
+        val bucketList = emptyList<String>()
+        resourceCache().s3buckets(bucketList)
+        bucketList.map { resourceCache().bucketRegion(it) }
+        val children = S3ServiceNode(projectRule.project).children
+        assertThat(children).allMatch { it is AwsExplorerEmptyNode }
+    }
+
+    @Test
+    fun errorLoadingBuckets() {
+        val children = S3ServiceNode(projectRule.project).children
```

Comment:
What causes this test to throw the exception?

Reply:
"IllegalStateException: No value found for ClientCachedBackedResource in mock" when it tried calling "LIST_BUCKETS"

Reply:
Change the test to be explicit in behavior. Put a failed future in the cache for example.